core/matrix/csr_lookup.hpp

```diff
@@ -262,15 +264,17 @@ struct device_sparsity_lookup {
             (static_cast<unsigned_index_type>(col) * hash_param) % hashmap_size;
         GKO_ASSERT(hashmap[hash] >= 0);
         GKO_ASSERT(hashmap[hash] < row_nnz);
-        while (local_cols[hashmap[hash]] != col) {
+        auto out_idx = hashmap[hash];
+        // linear probing with invalid_index sentinel to avoid infinite loop
+        while (out_idx >= 0 && local_cols[out_idx] != col) {
             hash++;
             if (hash >= hashmap_size) {
                 hash = 0;
             }
-            GKO_ASSERT(hashmap[hash] >= 0);
+            out_idx = hashmap[hash];
             GKO_ASSERT(hashmap[hash] < row_nnz);
         }
-        const auto out_idx = hashmap[hash];
+        // out_idx is either correct or invalid_index, the hashmap sentinel
```

Comment:
I am confused here. With this change, `lookup_hash_unsafe` is the same as `lookup_hash` (except for one extra assertion for `unsafe` -- the `>=0` -- not sure why?). So...`lookup_hash_unsafe` isn't really unsafe anymore? Why do we need any changes to `unsafe` at all, when in the factorization, we switch to use the safe lookup if requested? (`lookup[col]` rather than `lookup.lookup_unsafe(col)`)

Reply:
`lookup_hash_unsafe` is mostly equal to `lookup_hash` up to one assert as you mentioned.
in my mind. when you call `lookup_unsafe`, you will assume it always returns a valid value back without any checks.
when calling `lookup`, you will always checks whether it is valid or not.
I keep them because of the extra assert. unsafe should report the index is invalid in debug mode because it may not be checked in the code.
I want to change the unsafe lookup because the infinite loop leads debug difficult and this change does not affect the performance.
BTW, I do not think there's performance difference between safe and unsafe version, but I will keep this behavior first.
Then, we can see whether to use safe-only lookup.